### PR TITLE
[OMSICpp] Add OMSI Cpp library and  omsicpp code target template

### PR DESCRIPTION
**Pull request by niklwors**
Simulation runtime for FMU/OMSU simulation in C++.

 - Added OMSICpp build
   - Use CMake to build SimultaionRuntime/OMSICpp
 - Added OMSICpp in Simulation Runtime
   - The code targets cpp and cpp fmu using this runtime also now
   - Using OMSIBase library for base functionalities shared with OMSIC runtime
   - Wrapper for FMI 2.0 ModelExchange functions
     - Functions for continuous simulation of FMU/OMSU
     - Functions for event simulation of FMU/OMSU
     - Getter and Setter functions for FMU/OMSU
     - Logging and some debugging functionalities
     - Initialization and deallocation of FMU/OMSU
   - OMSICpp Simulation runtime which supports omsu/fmu 2.0 ME and  the old
     cpp code traget systems
 - Documentation with Doxygen
   - Doxyfile not included

Co-authored-by: AnHeuermann <andreas.heuermann@fh-bielefeld.de>
Co-authored-by: wibraun <wbraun@fh-bielefeld.de>

### CODE REVIEW

**Comment by sjoelund** [comment]
There seems to be additional problems:
* Calling `make omc` twice causes all of OMSICPP to be recompiled. **Not OK**
* You have dependencies like: `OMSICPP: CMinpack sundials antlr-copy $(MINGW_EXTRA_LIBS) fmil OMSI omc` which are very unsafe as the `fmil` dependency might now be called in parallel (there is a job that calls `make -j20 omc runtimeCPPinstall` and `omc` also includes the `fmil` dependency). Any common dependencies to `omc` need to be stated in the `all-runtimeCPPinstall` and `omc-and-runtimeCPPinstall` rules. For that OMSICPP rule, any dependency that is part of `omc` should be omitted and only `omc` should be a dependency. Which is also a little odd. What do you need from omc to compile runtimeCPPinstall? It should be split in the compiler-dependent part and compiler-independent part to improve compilation speed if it's just for that one file that calls compiler API functions.

Note that Jenkins can sometimes run on 20-core machines, so testing on a laptop is often not sufficient. I would suggest running the compilation several times on Jenkins and making sure you get asap and ripper a few of those times when you are done to make sure you don't randomly break the builds later on....

**Comment by niklwors** [comment]
@sjoelund  I adapted the make files to your proposal. "Calling make omc twice ..."  is still missing. Should this be implemented with the test command? What happens if a source file has been changed and you want to rebuild the runtime?

**Comment by sjoelund** [comment]
If a single file changes, only that file should be recompiled. But running make omc && make omc from the OpenModelica directory should be the almost the same speed as just make omc. Like all the other projects. Perhaps make sure you don't call cmake unless the CMakeList.txt files changed for example. Or don't make clean or remove directories before the make command starts.

File: OMCompiler/SimulationRuntime/OMSICpp/runtime/src/Core/System/OSUSystem.cpp [hunk 1]
```diff
@@ -533,9 +534,10 @@ bool OSUSystem::handleSystemEvents(bool* events)
       std::string error = std::string("fmi2GetEventIndicators failed with status ::") + std::string(log_str);
       throw ModelicaSimulationError(MODEL_EQ_SYSTEM, error);
     }
-
+    cout << dim zero = _dimZeroFunc;
     for (int i = 0; i < _dimZeroFunc; i++)
     {
+	    cout << "nein" << std:endl;
```
Comment (hkiel) [inline OMCompiler/SimulationRuntime/OMSICpp/runtime/src/Core/System/OSUSystem.cpp]: left over debugging code?

File: OMCompiler/Compiler/Template/CodegenCpp.tpl [hunk 2]
```diff
@@ -123,6 +136,7 @@ case SIMCODE(__) then
    <<
    <%generateHeaderIncludeString(simCode, &extraFuncs , &extraFuncsDecl, extraFuncsNamespace)%>
    <%additionalIncludes%>
+   /*Test*/
```
[the remark this answers is not in the document]
Comment (niklwors) [reply OMCompiler/Compiler/Template/CodegenCpp.tpl]: is removed

File: OMCompiler/Compiler/Template/CodegenOMSICpp.tpl [hunk 3]
```diff
@@ -0,0 +1,1062 @@
...
+
+template osuModelCppFile(SimCode simCode,Text& extraFuncs,Text& extraFuncsDecl,Text extraFuncsNamespace, String guid, String FMUVersion)
+ "Generates code for OSU target."
+::=
+match simCode
+case SIMCODE(modelInfo=MODELINFO(__)) then
+  let modelName = dotPath(modelInfo.name)
+  let modelShortName = lastIdentOfPath(modelInfo.name)
+  let modelLongName = System.stringReplace(modelName, ".", "_")
+  <<
```
[the remark this answers is not in the document]
Comment (niklwors) [reply OMCompiler/Compiler/Template/CodegenOMSICpp.tpl]: Yes this file is not needed  to generate. We can move this file in the runtime in a next Pull Request

File: OMCompiler/SimulationRuntime/OMSICpp/CMakeLists.txt [hunk 4]
```diff
@@ -816,87 +934,78 @@ if(UNIX AND Boost_INCLUDE_DIR STREQUAL "/usr/include")
...
-IF(USE_DGESV)
-  ADD_SUBDIRECTORY(Solver/Dgesv)
+#include path for cpp simulation kernel
+include_directories ("runtime/include")
+#include path for OpenModelica Simulation Interface (OMSI)
+include_directories ("../OMSI/include")
+
+
+
+#TODO: OMSIMerge verify changes
```
[the remark this answers is not in the document]
Comment (niklwors) [reply OMCompiler/SimulationRuntime/OMSICpp/CMakeLists.txt]: yes, it was verified

File: OMCompiler/Makefile.omdev.mingw [hunk 5]
```diff
@@ -97,7 +96,7 @@ include Makefile.common
 include Makefile.omsi.common
 
 # For Windows build static FMI lib!
-FMILIB_SHARED = OFF
+FMILIB_SHARED = ON
```
Comment (AnHeuermann) [inline OMCompiler/Makefile.omdev.mingw]: @adrpo There was probably a realy good reason for disabling FMI Library to be build as shared library on Windows systems. Do you kow why and if it's still needed?

File: OMCompiler/SimulationRuntime/OMSICpp/runtime/src/Solver/CVode/CVode.cpp [hunk 6]
```diff
@@ -484,8 +485,9 @@ void Cvode::CVodeCore()
 			throw ModelicaSimulationError(SOLVER, "CVodeGetLastStep failed. The cvode mem pointer is NULL");
 
 		//set completed step to system and check if terminate was called
-		if (_continuous_system->stepCompleted(_tCurrent))
-			_solverStatus = DONE;
+    /*Todo: replaced by stepCompleted, check for terminate has to be done*/
```
[the remark this answers is not in the document]
Comment (niklwors) [reply OMCompiler/SimulationRuntime/OMSICpp/runtime/src/Solver/CVode/CVode.cpp]: in the current version the terminate command is not supported, this has to be done in a subsequent PR

File: OMCompiler/SimulationRuntime/OMSICpp/runtime/src/Core/System/OSUSystem.cpp (hunk 1 above)
Comment (niklwors) [reply OMCompiler/SimulationRuntime/OMSICpp/runtime/src/Core/System/OSUSystem.cpp]: Is removed.

**Comment by sjoelund** [comment]
I wish this was tested on OSX before it was merged. It needs the fixes I made to the cppruntime to make OSX build again... Especially 3e60c14bfc729519ed7944cf7ab1f726323f9a04

**Comment by niklwors** [comment]
> 
> 
> I wish this was tested on OSX before it was merged. It needs the fixes I made to the cppruntime to make OSX build again... Especially [3e60c14](https://github.com/OpenModelica/OpenModelica/commit/3e60c14bfc729519ed7944cf7ab1f726323f9a04)

This should be already merged in the new cmake file OMCompiler\SimulationRuntime\OMSICpp\CMakeLists.txt

**Comment by sjoelund** [comment]
OK, it seems to be an identical copy-paste in 8633ec15ac094110f7810b99ff4ae12010f21801. Same (broken) lines were used in OMCompiler/SimulationRuntime/OMSI/solver/CMakeLists.txt.
You can see the logs in https://test.openmodelica.org/jenkins/job/periodic/job/OSX/job/master/5/consoleFull. And it's now possible to start a pull request to also build for OSX (https://test.openmodelica.org/jenkins/job/OpenModelica/view/change-requests/, select the PR to run and press Build Now, and check the OSX box).

File: OMCompiler/Makefile.omdev.mingw (hunk 5 above)
Comment (adrpo) [reply OMCompiler/Makefile.omdev.mingw]: I don't remember why I did this. You can try with it ON and see if is fine.

**Pull request merged**